lib: net_buf: support getting max used buf and packet

MaochenWang1 opened the pull request:
lib: net_buf: support getting max used buf and packet
When defined both CONFIG_NET_BUF_POOL_USAGE and CONFIG_MEM_SLAB_TRACE_MAX_UTILIZATION, support using 'net mem' cmd to get the maximum count of used buffers and net packets.

jhedberg commented:
Please separate the net_buf changes to a separate commit from any other net changes. The prefix to that commit should likely be "lib: net_buf: ..." now that the net_buf implementation has been moved out from the networking directories.

MaochenWang1 commented:
> Please move the net_buf changes to a separate commit from any other net changes. The prefix to that commit should likely be "lib: net_buf: ..." now that the net_buf implementation has been moved out from the networking directories.

updated, please review

jukkar commented on subsys/net/lib/shell/mem.c:
```diff
@@ -114,19 +114,25 @@ static int cmd_net_mem(const struct shell *sh, size_t argc, char *argv[])
 	PR("Network buffer pools:\n");
```
As @jhedberg mentioned earlier, please place the net-shell changes to separate commit (can be in the same PR).

MaochenWang1 replied:
updated, please review

MaochenWang1 committed:
lib: net_buf: support counting max used buf

Support counting the max used net_buf when CONFIG_NET_BUF_POOL_USAGE
is defined.

Signed-off-by: Maochen Wang <maochen.wang@nxp.com>

MaochenWang1 committed:
net: shell: mem: support getting max used buf and packet

When defined both CONFIG_NET_BUF_POOL_USAGE and
CONFIG_MEM_SLAB_TRACE_MAX_UTILIZATION, support using 'net mem' cmd
to get the maximum count of used buffers and net packets.

Signed-off-by: Maochen Wang <maochen.wang@nxp.com>

jhedberg commented on lib/net_buf/buf.c:
```diff
@@ -344,6 +344,8 @@ struct net_buf *net_buf_alloc_len(struct net_buf_pool *pool, size_t size,
 #if defined(CONFIG_NET_BUF_POOL_USAGE)
 	atomic_dec(&pool->avail_count);
 	__ASSERT_NO_MSG(atomic_get(&pool->avail_count) >= 0);
+	pool->max_used = MAX(pool->max_used,
+		pool->buf_count - atomic_get(&pool->avail_count));
```
minor: please fix the alignment here. It should be n tabs + 0-7 spaces so that the split line starts right after the ( on the previous line

MaochenWang1 replied:
updated, thanks

jhedberg commented on include/zephyr/net_buf.h:
```diff
@@ -1098,6 +1098,9 @@ struct net_buf_pool {
 	/** Total size of the pool. */
 	const uint16_t pool_size;
 
+	/** Maximum count of used buffers. */
+	uint16_t max_used;
```
One thing I started wondering: perhaps this member should also be behind `#if defined(CONFIG_NET_BUF_POOL_USAGE)`?

pdgendt replied:
It is (open some more context above)

jhedberg replied:
Right. Thanks.

MaochenWang1 commented:
Hi @jukkar please review
